core/amber/src/test/scala/edu/uci/ics/amber/engine/architecture/scheduling/DefaultCostEstimatorSpec.scala

```diff
@@ -0,0 +1,206 @@
+    workflowRuntimeStatistics.setDataProcessingTime(ULong.valueOf(200))
+    workflowRuntimeStatistics.setControlProcessingTime(ULong.valueOf(200))
+    workflowRuntimeStatistics
+  }
+
+  override protected def beforeAll(): Unit = {
+    initializeDBAndReplaceDSLContext()
+  }
+
+  "DefaultCostEstimator" should "use fallback method when no past statistics are available" in {
+    val workflow = buildWorkflow(
+      List(headerlessCsvOpDesc, keywordOpDesc, sink),
+      List(
+        LogicalLink(
+          headerlessCsvOpDesc.operatorIdentifier,
+          PortIdentity(0),
+          keywordOpDesc.operatorIdentifier,
+          PortIdentity(0)
+        ),
+        LogicalLink(
+          keywordOpDesc.operatorIdentifier,
+          PortIdentity(0),
+          sink.operatorIdentifier,
+          PortIdentity(0)
+        )
+      ),
+      new WorkflowContext()
+    )
+
+    val costEstimator = new DefaultCostEstimator(
+      workflow.context,
+      CONTROLLER
+    )
+
+    val region = Region(
+      id = RegionIdentity(0),
+      physicalOps = workflow.physicalPlan.operators,
+      physicalLinks = workflow.physicalPlan.links
+    )
+
+    val costOfRegion = costEstimator.estimate(region, 1)
+
+    assert(costOfRegion == 0)
+  }
+
+  "DefaultCostEstimator" should "use the latest successful execution to estimate cost when available" in {
+    val workflow = buildWorkflow(
+      List(headerlessCsvOpDesc, keywordOpDesc, sink),
+      List(
+        LogicalLink(
+          headerlessCsvOpDesc.operatorIdentifier,
+          PortIdentity(0),
+          keywordOpDesc.operatorIdentifier,
+          PortIdentity(0)
+        ),
+        LogicalLink(
+          keywordOpDesc.operatorIdentifier,
+          PortIdentity(0),
+          sink.operatorIdentifier,
+          PortIdentity(0)
+        )
+      ),
+      new WorkflowContext()
+    )
+
+    val userDao = new UserDao(getDSLContext.configuration())
+    val workflowDao = new WorkflowDao(getDSLContext.configuration())
+    val workflowExecutionsDao = new WorkflowExecutionsDao(getDSLContext.configuration())
+    val workflowVersionDao = new WorkflowVersionDao(getDSLContext.configuration())
+    val workflowRuntimeStatisticsDao =
+      new WorkflowRuntimeStatisticsDao(getDSLContext.configuration())
+
+    userDao.insert(testUser)
+    workflowDao.insert(testWorkflowEntry)
+    workflowVersionDao.insert(testWorkflowVersionEntry)
+    workflowExecutionsDao.insert(testWorkflowExecutionEntry)
+    workflowRuntimeStatisticsDao.insert(headerlessCsvOpStatisticsEntry)
+    workflowRuntimeStatisticsDao.insert(keywordOpDescStatisticsEntry)
+    workflowRuntimeStatisticsDao.insert(sinkOpDescStatisticsEntry)
+
+    val costEstimator = new DefaultCostEstimator(
+      workflow.context,
+      CONTROLLER
+    )
+
+    val region = Region(
+      id = RegionIdentity(0),
+      physicalOps = workflow.physicalPlan.operators,
+      physicalLinks = workflow.physicalPlan.links
+    )
```

Comment:
can we add a test case where the region only contains a sub-dag of the workflow? just to make sure it can extract region costs correctly.

Reply:
Added one more test case.